### PR TITLE
Remove pydantic import warning for Langchain >= 0.3.0

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/13190?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13190/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13190
```

</p>
</details>

### What changes are proposed in this pull request?

Creating a trace with Langchain `0.3.0` emits a warning for the usage of old pydantic import

Screenshot 2024-09-19 at 10 56 04

This PR removes this warning by updating the import statement.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

No warning is issued when this patch is installed:

Screenshot 2024-09-19 at 10 51 23

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/langchain/runnables.py [hunk 1]
```diff
@@ -262,14 +262,11 @@ def _save_internal_runnables(runnable, path, loader_fn, persist_dir):
         if hasattr(runnable, "save"):
             runnable.save(model_path)
         elif hasattr(runnable, "dict"):
-            try:
```
Comment (B-Step62) [inline mlflow/langchain/runnables.py:265]: This is unrelated change, but during debugging the issue, I found the `Cannot save runnable without `save` method` is not always correct. It's often triggered by different errors, for example, a loading issue during `_load_model_from_config`.

**Comment by B-Step62** [comment]
Cross version test failure in `langchain / models / 0.3.0` should not be related to this change.... need to tackle them anyway tho😭

File: mlflow/langchain/runnables.py (hunk 1 above)
Comment (daniellok-db) [reply mlflow/langchain/runnables.py:265]: is the normal error message in this case informative enough?

File: mlflow/tracing/utils/__init__.py [hunk 2]
```diff
@@ -47,15 +47,19 @@ class TraceJSONEncoder(json.JSONEncoder):
...
     def default(self, obj):
         try:
-            # LangChain does some trick to keep both Pydantic 1.x and 2.x support, so checking
+            import langchain
+
+            # LangChain < 0.3.0 does some trick to support Pydantic 1.x and 2.x, so checking
             # type with installed Pydantic version might not work for some models.
             # https://github.com/langchain-ai/langchain/blob/b66a4f48fa5656871c3e849f7e1790dfb5a4c56b/libs/core/langchain_core/pydantic_v1/__init__.py#L7
-            from langchain_core.pydantic_v1 import BaseModel as LangChainBaseModel
+            if Version(langchain.__version__) < Version("0.3.0"):
```
Comment (daniellok-db) [inline mlflow/tracing/utils/__init__.py:55]: just to clarify, we don't need to do this >= 0.3.0 because it's only pydantic v2?

**Review by daniellok-db — APPROVED** [review]
lgtm overall

File: mlflow/langchain/runnables.py (hunk 1 above)
Comment (B-Step62) [reply mlflow/langchain/runnables.py:265]: I think so. The exception here is not exposed directly, we add some prefix to it [here](https://github.com/mlflow/mlflow/blob/f9d760b3ba553762a5a9b3b6c2f901297ac1a0a8/mlflow/langchain/runnables.py#L363-L367). So the actual message shown to user is like
```
Failed to save runnable sequence:  {0: ChatDatabricks ..} -- Actual exception
```

File: mlflow/tracing/utils/__init__.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/tracing/utils/__init__.py:55]: Yes, their entire code base now imports the actual pydantic `BaseModel`.
https://github.com/search?q=repo%3Alangchain-ai%2Flangchain+%22import+BaseModel%22&type=code (except some stale examples)

**Comment by BenWilson2** [comment]
@B-Step62 let's force merge this as the LC 0.3.0 is unrelated and this is a good fix of an annoying warning.

**Comment by B-Step62** [comment]
> @B-Step62 let's force merge this as the LC 0.3.0 is unrelated and this is a good fix of an annoying warning.

Oh yea I totally forgot about this one😅

**Pull request merged**